src/test/java/org/springframework/data/elasticsearch/junit/jupiter/SpringDataElasticsearchExtension.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.elasticsearch.junit.jupiter;
+
+import java.util.List;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.test.context.ContextConfigurationAttributes;
+import org.springframework.test.context.ContextCustomizer;
+import org.springframework.test.context.ContextCustomizerFactory;
+import org.springframework.test.context.MergedContextConfiguration;
+
+/**
+ * This extension class check in the {@link #beforeAll(ExtensionContext)} call if there is already a Elasticsearch
+ * cluster connection defined in the root store. If no, the connection to the cluster is defined according to the
+ * configuration, starting a local node if necessary. The connection is stored and will be closed when the store is
+ * shutdown at the end of all tests
+ *
+ * A ParameterResolver is implemented which enables resolving the
+ * ClusterConnectionInfo to test methods, and if a Spring context is used, the ClusterConnectionInfo can be autowired.
+ *
+ * @author Peter-Josef Meisch
+ */
+public class SpringDataElasticsearchExtension
+		implements BeforeAllCallback, ParameterResolver, ContextCustomizerFactory {
+
+	public static final String SPRING_DATA_ELASTICSEARCH_TEST_CLUSTER_URL = "SPRING_DATA_ELASTICSEARCH_TEST_CLUSTER_URL";
+	private static final Logger LOGGER = LoggerFactory.getLogger(SpringDataElasticsearchExtension.class);
+
+	private static final String STORE_KEY = ClusterConnection.class.getSimpleName();
+
+	private static final Lock initLock = new ReentrantLock();
+
+	@Override
+	public void beforeAll(ExtensionContext extensionContext) throws Exception {
+		initLock.lock();
+		try {
+			ExtensionContext rootContext = extensionContext.getRoot();
```

Comment:
In this method we can guarantee single-threaded behavior, therefore I'd suggest looking up `ClusterConnection`  and storing it in the `ExtensionContext.store`, too. With migrating `resolveParameter(…)` to use the extension store, we no longer depend on threading for the parameter lookup. The only integration that needs to worry about `Thread` affinity is how we pass `ClusterConnection` into the Spring context.